crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+            .lookup_path(&normalized.with_trailing_slash())
+            .is_ok_and(|zipfile| zipfile.is_dir())
+    }
+
+    fn is_file(&self, path: &FileSystemPath) -> bool {
+        if path.as_str().ends_with('/') {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalize_vendored_path(path))
+            .is_ok_and(|zipfile| zipfile.is_file())
+    }
+
+    fn metadata(&self, path: &FileSystemPath) -> Result<Metadata> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+
+        let is_dir = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => zipfile.is_dir(),
+            Err(err) => {
+                if normalized.has_trailing_slash() {
+                    return Err(err);
+                }
+                let mut archive = self.vendored_archive();
+                let lookup_result = archive.lookup_path(&normalized.with_trailing_slash());
+                if let Ok(zipfile) = lookup_result {
+                    zipfile.is_dir()
+                } else {
+                    return Err(err);
+                }
+            }
+        };
+
+        let file_type = if is_dir {
+            FileType::Directory
+        } else {
+            FileType::File
+        };
+
+        Ok(Metadata {
+            revision: FileRevision::zero(),
+            permissions: Some(Self::READ_ONLY_PERMISSIONS),
+            file_type,
+        })
+    }
+
+    fn read(&self, path: &FileSystemPath) -> Result<String> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+        let lookup_error = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => return Self::read_zipfile(zipfile),
+            Err(err) => err,
+        };
+        if normalized.has_trailing_slash() {
+            return Err(lookup_error);
+        }
+        let mut archive = self.vendored_archive();
+        let zipfile = archive.lookup_path(&normalized.with_trailing_slash())?;
+        Self::read_zipfile(zipfile)
+    }
+}
+
+/// Newtype wrapper around a ZipArchive.
+///
+/// Since ZipArchives are generally cheap to clone (for now), this should be too.
+#[derive(Debug, Clone)]
+struct VendoredZipArchive(ZipArchive<Cursor<Arc<[u8]>>>);
+
+impl VendoredZipArchive {
+    fn new(vendored_bytes: Arc<[u8]>) -> ZipResult<Self> {
+        Ok(Self(ZipArchive::new(Cursor::new(vendored_bytes))?))
+    }
+
+    fn lookup_path(&mut self, path: &NormalizedVendoredPath) -> Result<ZipFile> {
+        Ok(self.0.by_name(path.as_str())?)
+    }
+}
+
+/// A path that has been normalized via the `normalize_vendored_path` function.
+///
+/// Trailing slashes are normalized away by `camino::Utf8PathBuf`s,
+/// but trailing slashes are crucial for distinguishing between
+/// files and directories inside zip archives.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct NormalizedVendoredPath(String);
```

Comment:
I think we could combine this with the `VendoredFilePath` and `VendoredFilePathBuf` where we require the use of `VendoredFilePathBuf::new` (that normalizes the string).